File: testSuite/test-Millennium-tree-builder.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;
use lib $ENV{'GALACTICUS_EXEC_PATH'}."/perl";
use File::Which;

# Take trees from the Millennium Database and process into both Galacticus and IRATE formats.
# Andrew Benson (12-October-2012)

# Create a file in Galacticus format.
system("cd ..; ./Galacticus.exe testSuite/parameters/mergerTreeFileBuildMillennium.xml");
die("FAILED: failed to make Galacticus-format merger tree file from Millennium database output")
    unless ( $? == 0 );

# Extract a single tree from the file we just created.
system("cd ..; ./scripts/aux/extractSingleTree.py testSuite/outputs/millenniumTestTreesGLC.hdf5 testSuite/outputs/millenniumTestTreesGLCSingle.hdf5 79000000");
die("FAILED: failed to extract a single merger tree")
    unless ( $? == 0 );

# Run the single tree to verify that the file is valid.
system("cd ..; ./Galacticus.exe testSuite/parameters/extractSingleTreeRun.xml");
die("FAILED: failed to run single merger tree")
    unless ( $? == 0 );

# Create a file in IRATE format.
system("cd ..; ./Galacticus.exe testSuite/parameters/mergerTreeFileBuildMillenniumIRATE.xml");
die("FAILED: failed to make IRATE-format merger tree file from Millennium database output")
    unless ( $? == 0 );

# Validate the IRATE format file.
my $validator = &File::Which::which('iratevalidate');
if ( $validator ) {
    system("cd ..; iratevalidate testSuite/outputs/millenniumTestTreesIRATE.hdf5");
    die("FAILED: IRATE-format file generated from Millennium database output did not validate")
	unless ( $? == 0 );
} else {
    print "SKIPPED: iratevalidate is not installed - validation of IRATE-format file will be skipped\n";
}

exit;

File: testSuite/test-Millennium-tree-builder.py
#!/usr/bin/env python3
import subprocess
import sys
import shutil

# Take trees from the Millennium Database and process into both Galacticus and IRATE formats.
# Andrew Benson (ported to Python)

# Create a file in Galacticus format.
status = subprocess.run("cd ..; ./Galacticus.exe testSuite/parameters/mergerTreeFileBuildMillennium.xml", shell=True)
if status.returncode != 0:
    print("FAILED: failed to make Galacticus-format merger tree file from Millennium database output")
    sys.exit(0)

# Extract a single tree from the file we just created.
status = subprocess.run("cd ..; ./scripts/aux/extractSingleTree.py testSuite/outputs/millenniumTestTreesGLC.hdf5 testSuite/outputs/millenniumTestTreesGLCSingle.hdf5 79000000", shell=True)
if status.returncode != 0:
    print("FAILED: failed to extract a single merger tree")
    sys.exit(0)

# Run the single tree to verify that the file is valid.
status = subprocess.run("cd ..; ./Galacticus.exe testSuite/parameters/extractSingleTreeRun.xml", shell=True)
if status.returncode != 0:
    print("FAILED: failed to run single merger tree")
    sys.exit(0)

# Create a file in IRATE format.
status = subprocess.run("cd ..; ./Galacticus.exe testSuite/parameters/mergerTreeFileBuildMillenniumIRATE.xml", shell=True)
if status.returncode != 0:
    print("FAILED: failed to make IRATE-format merger tree file from Millennium database output")
    sys.exit(0)

# Validate the IRATE format file.
validator = shutil.which("iratevalidate")
if validator:
    status = subprocess.run("cd ..; iratevalidate testSuite/outputs/millenniumTestTreesIRATE.hdf5", shell=True)
    if status.returncode != 0:
        print("FAILED: IRATE-format file generated from Millennium database output did not validate")
        sys.exit(0)
else:
    print("SKIPPED: iratevalidate is not installed - validation of IRATE-format file will be skipped")
